File: ge-develop/api/python/llm_datadist_v1/llm_datadist.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# -----------------------------------------------------------------------------------------------------------
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of 
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, 
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
from typing import Any, Dict, List, Tuple, Optional, Union
import atexit
from llm_datadist_v1 import llm_wrapper

try:
    from llm_datadist.utils.utils import check_isinstance, check_uint64, check_int32
except ModuleNotFoundError:
    from llm_datadist_v1.utils.utils import check_isinstance, check_uint64, check_int32

try:
    from llm_datadist.utils import log
except ModuleNotFoundError:
    from llm_datadist_v1.utils import log

try:
    from llm_datadist.status import (code_2_status, handle_llm_status, raise_if_false, LLMStatusCode,
                                    LLMException)
except ModuleNotFoundError:
    from llm_datadist_v1.status import (code_2_status, handle_llm_status, raise_if_false, LLMStatusCode,
                                    LLMException)

try:
    from llm_datadist.configs import LLMRole, LLMClusterInfo
except ModuleNotFoundError:
    from llm_datadist_v1.configs import LLMRole, LLMClusterInfo

try:
    from llm_datadist.v2.config import EngineConfig
except ModuleNotFoundError:
    from llm_datadist_v1.config import EngineConfig

try:
    from llm_datadist.v2.llm_utils import parse_listen_ip_info
except ModuleNotFoundError:
    from llm_datadist_v1.llm_utils import parse_listen_ip_info

from llm_datadist_v1.kv_cache_manager import KvCacheManager

__all__ = ['LLMDataDist', 'KvCacheManager']

class LLMDataDist(object):
    llm_engine_instance = None
    """
    LLMDataDist

    Args:
        role: role of LLMDataDist
        cluster_id: cluster_id of LLMDataDist
    """

    def __init__(self, role: LLMRole, cluster_id: int):
        check_isinstance("role", role, LLMRole)
        self._kv_cache_manager = None
        self._cache_manager = None
        self._role = role
        check_uint64("cluster_id", cluster_id)
        self._cluster_id = cluster_id
        self._llm_datadist = None
        self._engine_config = None
        self._is_initialized = False
        self._engine_options: Dict[str, str] = {}
        self._enable_cache_mgr = False
        self._enable_free_comm = False
        self._enable_local_comm_res = False

    def _check_flow_graph_max_size(self, options: Dict[str, str]) -> None:
        mem_utilization = float(options.get("llm.MemoryUtilization", "0.95"))
        value = options.get("ge.flowGraphMemMaxSize", None)
        if value is None:
            return
        check_isinstance('ge.flowGraphMemMaxSize', value, str)
        raise_if_false(len(value.split(",")) == 1, "ge.flowGraphMemMaxSize only support one mem pool in llm datadist")
        raise_if_false(value.isdigit(), "ge.flowGraphMemMaxSize must be digit")

    def init(self, options: Dict[str, str]) -> None:
        """
        初始化LLM Engine

        Args:
            options: Engine相关options
        """
        if self._is_initialized:
            return
        raise_if_false(LLMDataDist.llm_engine_instance is None, 'Cannot init multiple LLM engines',
                       status_code=LLMStatusCode.LLM_FAILED)
        check_isinstance("options", options, dict)
        self._check_flow_graph_max_size(options)
        self._engine_options = options
        self._engine_options['llm.Role'] = self._role_to_str(self._role)
        self._enable_local_comm_res = "llm.LocalCommRes" in options
        if self._enable_local_comm_res and "llm.EnableCacheManager" not in options:
            self._engine_options["llm.EnableCacheManager"] = "1"
        if self._enable_local_comm_res and "llm.EnableRemoteCacheAccessible" not in options:
            self._engine_options["llm.EnableRemoteCacheAccessible"] = "1"
        self._enable_cache_mgr = (
            "llm.EnableCacheManager" in self._engine_options and
            self._engine_options["llm.EnableCacheManager"] == "1"
        )
        log.info('options = %s', self._engine_options)
        self._llm_datadist = llm_wrapper
        EngineConfig.gen_cluster_info_if_not_exist(self._cluster_id, self._role, self._engine_options)
        ret = self._llm_datadist.initialize(self._cluster_id, self._engine_options)
        handle_llm_status(ret, '[LLMDataDist.init]', f'Failed to initialize llm datadist, options = {options}')
        self._kv_cache_manager = KvCacheManager(self._llm_datadist, self._role)
        LLMDataDist.llm_engine_instance = self
        self._is_initialized = True

    def _check_is_cache_mgr_mode(self, func_name):
        raise_if_false(self._enable_cache_mgr,
                       '{0} is not supported when llm.EnableCacheManager is not configured.',
                       func_name)

    def _check_is_not_cache_mgr_mode(self, func_name):
        raise_if_false(not self._enable_cache_mgr,
                       '{0} is not supported when llm.EnableCacheManager is configured.',
                       func_name)

    def finalize(self) -> None:
        """
        释放LLM Engine相关资源
        """
        if not self._is_initialized:
            return
        self._llm_datadist.finalize()
        if self._kv_cache_manager is not None:
            self._kv_cache_manager._initialized = False
        self._is_initialized = False
        LLMDataDist.llm_engine_instance = None

    def _cluster_config(self):
        if self._engine_config is None:
            self._engine_config = EngineConfig.from_engine_options(self._role == LLMRole.PROMPT, self._engine_options)
        return self._engine_config.cluster_config


    def check_link_status(self, remote_cluster_id: int):
        self._check_is_inited()
        self._check_is_not_cache_mgr_mode('check_link_status')
        check_uint64("remote_cluster_id", remote_cluster_id)
        ret = self._llm_datadist.check_link_status(remote_cluster_id)
        handle_llm_status(ret, '[check_link_status]', f"remote_cluster_id is {remote_cluster_id}")
        log.info('[check_link_status] success')


    def link_clusters(self, clusters: Union[List[LLMClusterInfo], Tuple[LLMClusterInfo]], timeout=3000):
        self._check_is_inited()
        check_int32("timeout", timeout)
        raise_if_false(timeout > 0, "Param timeout should be greater than 0.")
        check_isinstance("clusters", clusters, [list, tuple], LLMClusterInfo)
        cluster_list = [(cluster.remote_cluster_id,
                         0,
                         cluster.local_ip_info_list,
                         cluster.remote_ip_info_list) for cluster in clusters]
        ret, rets = self._llm_datadist.link_clusters(cluster_list, timeout)
        return code_2_status(ret), [code_2_status(cluster_ret) for cluster_ret in rets]

    def unlink_clusters(self, clusters: Union[List[LLMClusterInfo], Tuple[LLMClusterInfo]], timeout=3000, force=False):
        self._check_is_inited()
        self._check_is_not_cache_mgr_mode('unlink_clusters')
        check_int32("timeout", timeout)
        raise_if_false(timeout > 0, "Param timeout should be greater than 0.")
        check_isinstance("clusters", clusters, [list, tuple], LLMClusterInfo)
        check_isinstance("force", force, bool)
        cluster_list = [(cluster.remote_cluster_id,
                         0,
                         cluster.local_ip_info_list,
                         cluster.remote_ip_info_list) for cluster in clusters]
        ret, rets = self._llm_datadist.unlink_clusters(cluster_list, timeout, force)
        return code_2_status(ret), [code_2_status(cluster_ret) for cluster_ret in rets]

    def switch_role(self, role: LLMRole, switch_options: Optional[Dict[str, str]] = None):
        self._check_is_inited()
        check_isinstance('role', role, LLMRole)
        raise_if_false(self._role != role, f'role not changed, role = {role.name}')
        role_str = self._role_to_str(role)
        log.info(f'[switch_role] [{self._role.name}->{role.name}] start, switch_options = {switch_options}')
        check_isinstance('switch_options', switch_options, dict)
        options = switch_options.copy() if switch_options is not None else {}
        if role == LLMRole.PROMPT:
            raise_if_false('llm.listenIpInfo' in switch_options,
                           'Failed to switch to Prompt, option "llm.listenIpInfo" was specified')
            listen_ip_info = switch_options['llm.listenIpInfo']
            ip, port = EngineConfig.parse_listen_ip_info(listen_ip_info)
            options['llm.ListenIp'] = str(ip)
            options['llm.ListenPort'] = str(port)
        ret = self._llm_datadist.switch_role(role_str, options)
        handle_llm_status(ret,
                          '[LLMEngine.switch_role]',
                          f'Failed to switch role, role = {role}, options = {options}')
        self._kv_cache_manager._switch_role(role)
        log.info(f'[switch_role] [{self._role.name}->{role.name}] success')
        self._role = role

    @staticmethod
    def _role_to_str(role: LLMRole) -> str:
        role_mapping = {
            LLMRole.PROMPT: 'Prompt',
            LLMRole.DECODER: 'Decoder',
            LLMRole.MIX: 'Mix',
        }
        return role_mapping[role]

    def _check_is_inited(self):
        if not self._is_initialized:
            raise RuntimeError('llm datadist is not initialized')

    @property
    def kv_cache_manager(self) -> KvCacheManager:
        """
        获取KvCacheManager

        Returns:
            KvCacheManager
        """
        self._check_is_inited()
        self._check_is_not_cache_mgr_mode('kv_cache_manager')
        return self._kv_cache_manager

    @property
    def cluster_id(self):
        return self._cluster_id


def _shutdown_handler():
    if LLMDataDist.llm_engine_instance is not None:
        log.info('[shutdown_handler] finalize llm datadist')
        try:
            LLMDataDist.llm_engine_instance.finalize()
        except LLMException as e:
            log.warn(f'error occurred while finalize llm datadist: {e} '
                     f'may cause by already finalized by another framework')


atexit.register(_shutdown_handler)

File: ge-develop/dflow/llm_datadist/v1/llm_datadist.cc
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of 
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, 
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "llm_datadist.h"
#include "common/mem_utils.h"
#include "common/llm_flow_service.h"
#include "common/cache_engine.h"
#include "common/llm_common.h"
#include "common/llm_checker.h"
#include "common/llm_scope_guard.h"

namespace llm {
class LLMDataDist::Impl {
 public:
  explicit Impl(const uint64_t cluster_id) : cluster_id_(cluster_id) {}
  ~Impl() {
    if (is_initialized_) {
      LLMDataDistFinalize();
    }
  }

  ge::Status LLMDataDistInitialize(const std::map<ge::AscendString, ge::AscendString> &options);

  void LLMDataDistFinalize();

  bool IsInitialized() const;

  ge::Status CheckLinkStatus(uint64_t remote_cluster_id);

  ge::Status LinkClusters(const std::vector<ClusterInfo> &clusters, std::vector<ge::Status> &rets,
                          const int32_t timeout);

  ge::Status UnlinkClusters(const std::vector<ClusterInfo> &clusters, std::vector<ge::Status> &rets,
                            const int32_t timeout, bool force_flag = false);

  ge::Status AllocateCache(const CacheDesc &cache_desc,
                           Cache &cache,
                           const std::vector<CacheKey> &cache_keys);

  ge::Status DeallocateCache(int64_t cache_id);

  ge::Status RemoveCacheKey(const CacheKey &cache_key);

  ge::Status PullCache(int64_t cache_id, const CacheKey &cache_key, const PullCacheParam &pull_cache_param);

  ge::Status PullBlocks(int64_t cache_id, const CacheKey &cache_key, const PullCacheParam &pull_cache_param);

  ge::Status TransferCache(uint64_t task_id, const TransferCacheConfig &transfer_cache_config,
                           const TransferBlockConfig &transfer_block_config);

  ge::Status CopyCache(const CacheEntry &src_cache_entry, const CacheEntry &dst_cache_entry,
                       const CopyCacheParam &copy_cache_param, const std::vector<int32_t> &device_ids);

  ge::Status CopyCache(const CopyCacheParam &copy_cache_param);

  ge::Status GetCacheTensors(int64_t cache_id, std::vector<ge::Tensor> &outputs, int32_t tensor_index);

  ge::Status SwitchRole(const std::string &role, const std::map<std::string, std::string> &options);

  ge::Status SwapBlocks(const Cache &src, const Cache &dst, const uint64_t block_size, const uint32_t type,
                        const std::vector<std::pair<int64_t, int64_t>> &block_mapping);

 private:
  ge::Status InitFlowService(const std::map<ge::AscendString, ge::AscendString> &options);

  uint64_t cluster_id_;
  std::string role_;
  bool enable_switch_role_ = false;
  std::atomic<bool> is_initialized_{false};
  std::vector<int32_t> device_ids_;

  uint64_t model_id_ = 0U;
  std::mutex mutex_;
  std::unique_ptr<LlmFlowService> llm_flow_service_;
  std::unique_ptr<CacheEngine> cache_engine_;
};

ge::Status LLMDataDist::Impl::LLMDataDistInitialize(const std::map<ge::AscendString, ge::AscendString> &options) {
  LLMLOGI("LLMDataDist initialize cluster id:%lu", cluster_id_);
  std::lock_guard<std::mutex> lk(mutex_);
  if (!is_initialized_) {
    auto iter = options.find(LLM_OPTION_ROLE);
    LLM_ASSERT_TRUE(iter != options.end(), "[LLMDataDist] option:%s not found.", LLM_OPTION_ROLE);
    role_ = iter->second.GetString();
    LLMLOGI("LLMDataDist role:%s.", role_.c_str());
    LLM_ASSERT_TRUE((role_ == kDecoder) || (role_ == kPrompt) || (role_ == kMix), "Invalid role: %s", role_.c_str());
    const auto ret = InitFlowService(options);
    if (ret != ge::SUCCESS) {
      if (llm_flow_service_ != nullptr) {
        llm_flow_service_->Finalize();
      }
      LLMLOGE(ret, "[Manager] initialize failed.");
      return ret;
    }
    is_initialized_ = true;
    LLMLOGI("[LLMDataDist] role:[%s] cluster:[%lu] init success", role_.c_str(), cluster_id_);
  }
  return ge::SUCCESS;
}

void LLMDataDist::Impl::LLMDataDistFinalize() {
  std::lock_guard<std::mutex> lk(mutex_);
  if (!is_initialized_.load()) {
    LLMLOGW("[LLMDataDist] cluster:%lu is not initialized", cluster_id_);
    return;
  }
  llm_flow_service_->Finalize();
  is_initialized_.store(false);
  LLMLOGI("LLMDataDist Finalize end.");
}


bool LLMDataDist::Impl::IsInitialized() const {
  return is_initialized_.load(std::memory_order::memory_order_relaxed);
}

ge::Status LLMDataDist::Impl::CheckLinkStatus(uint64_t remote_cluster_id){
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(), ge::FAILED, "Llm datadist of cluster:%lu is not initialized.",
                         cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(role_ == kDecoder, ge::LLM_PARAM_INVALID,
                         "[LLMDataDist] CheckLinkStatus can only be used by Decoder, current_role is %s, ",
                         role_.c_str());
  LLM_CHK_STATUS_RET(llm_flow_service_->CheckLinkStatus(remote_cluster_id));
  return ge::SUCCESS;
}

ge::Status LLMDataDist::Impl::LinkClusters(const std::vector<ClusterInfo> &clusters,
                                           std::vector<ge::Status> &rets, const int32_t timeout) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(), ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(role_ == kDecoder,
                         ge::LLM_PARAM_INVALID,
                         "[LLMDataDist] linkClusters can only be used by Decoder, current_role is %s, ", role_.c_str());
  rets.clear();
  LLM_CHK_STATUS_RET(llm_flow_service_->LinkClusters(clusters, rets, timeout));
  return ge::SUCCESS;
}

ge::Status LLMDataDist::Impl::UnlinkClusters(const std::vector<ClusterInfo> &clusters,
                                             std::vector<ge::Status> &rets, const int32_t timeout, bool force_flag) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  rets.clear();
  LLM_CHK_STATUS_RET(llm_flow_service_->UnlinkClusters(clusters, rets, timeout, force_flag), "Unlink clusters failed");
  return ge::SUCCESS;
}

ge::Status LLMDataDist::Impl::AllocateCache(const CacheDesc &cache_desc,
                                            Cache &cache,
                                            const std::vector<CacheKey> &cache_keys) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  return cache_engine_->Allocate(cache_desc, cache_keys, cache);
}

ge::Status LLMDataDist::Impl::DeallocateCache(int64_t cache_id) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  return cache_engine_->Deallocate(cache_id);
}

ge::Status LLMDataDist::Impl::PullCache(int64_t cache_id,
                                        const CacheKey &cache_key,
                                        const PullCacheParam &pull_cache_param) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(role_ == kDecoder, ge::LLM_PARAM_INVALID,
                         "[PullKvCache] is only supported by Decoder, current_role is %s.",
                         role_.c_str());
  return cache_engine_->PullCache(cache_id, cache_key, pull_cache_param);
}

ge::Status LLMDataDist::Impl::PullBlocks(int64_t cache_id,
                                         const CacheKey &cache_key,
                                         const PullCacheParam &pull_cache_param) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(role_ == kDecoder, ge::LLM_PARAM_INVALID,
                         "[PullKvBlocks] is only supported by Decoder, current_role is %s.",
                         role_.c_str());
  LLM_CHK_BOOL_RET_STATUS((!pull_cache_param.prompt_blocks.empty()),
                         ge::LLM_PARAM_INVALID,
                         "src_blocks is empty, pull from non-block cache is not supported yet");
  LLM_CHK_BOOL_RET_STATUS(pull_cache_param.prompt_blocks.size() == pull_cache_param.decoder_blocks.size(),
                         ge::LLM_PARAM_INVALID,
                         "number of src_blocks (%zu) mismatches that of dst_blocks (%zu)",
                         pull_cache_param.prompt_blocks.size(), pull_cache_param.decoder_blocks.size());
  LLM_CHK_BOOL_RET_STATUS(cache_key.prompt_batch_index == 0U,
                         ge::LLM_PARAM_INVALID,
                         "invalid cache_key.prompt_batch_index (%lu), only 0 is supported in pull block",
                         cache_key.prompt_batch_index);
  return cache_engine_->PullCache(cache_id, cache_key, pull_cache_param);
}

ge::Status LLMDataDist::Impl::TransferCache(uint64_t task_id, const TransferCacheConfig &transfer_cache_config,
                                            const TransferBlockConfig &transfer_block_config) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(device_ids_.size() <= 1U, ge::LLM_FEATURE_NOT_ENABLED,
                        "Transfer cache is not supported when device num bigger than one.");
  LLM_CHK_BOOL_RET_STATUS(role_ == kPrompt, ge::LLM_PARAM_INVALID,
                         "Transfer cache is not supported by %s", role_.c_str());
  return cache_engine_->TransferCache(task_id, transfer_cache_config, transfer_block_config);
}

ge::Status LLMDataDist::Impl::CopyCache(const CopyCacheParam &copy_cache_param) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  return cache_engine_->CopyCache(copy_cache_param);
}

ge::Status LLMDataDist::Impl::CopyCache(const CacheEntry &src_cache_entry,
                                        const CacheEntry &dst_cache_entry,
                                        const CopyCacheParam &copy_cache_param,
                                        const std::vector<int32_t> &device_ids) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(src_cache_entry.cache_addrs.size() % device_ids_.size() == 0, ge::LLM_PARAM_INVALID,
                         "Param invalid, tensor_addrs size:%zu of src cache should be a multiple of device num:%zu.",
                         src_cache_entry.cache_addrs.size(), device_ids_.size());
  LLM_CHK_BOOL_RET_STATUS(dst_cache_entry.cache_addrs.size() % device_ids_.size() == 0, ge::LLM_PARAM_INVALID,
                         "Param invalid, tensor_addrs size:%zu of dst cache should be a multiple of device num:%zu.",
                         dst_cache_entry.cache_addrs.size(), device_ids_.size());
  return llm_flow_service_->CopyCache(src_cache_entry, dst_cache_entry, copy_cache_param, device_ids);
}

ge::Status LLMDataDist::Impl::RemoveCacheKey(const CacheKey &cache_key) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  return cache_engine_->RemoveCacheKey(cache_key);
}

ge::Status LLMDataDist::Impl::GetCacheTensors(int64_t cache_id,
                                              std::vector<ge::Tensor> &outputs,
                                              int32_t tensor_index) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed),
                         ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  return cache_engine_->GetCacheTensors(cache_id, outputs, tensor_index);
}

ge::Status LLMDataDist::Impl::InitFlowService(const std::map<ge::AscendString, ge::AscendString> &options) {
  LLM_CHK_STATUS_RET(LLMUtils::ParseFlag(LLM_OPTION_ENABLE_SWITCH_ROLE, options, enable_switch_role_));
  llm_flow_service_ = llm::MakeUnique<LlmFlowService>(role_, enable_switch_role_, cluster_id_);
  LLM_CHECK_NOTNULL(llm_flow_service_);
  LLM_CHK_STATUS_RET(LLMUtils::ParseDeviceId(options, device_ids_, ge::OPTION_EXEC_DEVICE_ID),
                    "Failed to get device id");
  if (device_ids_.empty()) {
    device_ids_.emplace_back(0U);
  }
  LLM_CHK_STATUS_RET(llm_flow_service_->Initialize(options, device_ids_), "Failed to initialize LLM flow service");

  cache_engine_ = llm::MakeUnique<CacheEngine>(llm_flow_service_.get());
  LLM_CHECK_NOTNULL(cache_engine_);

  LLMLOGI("Initialize LLM flow service successfully");
  return ge::SUCCESS;
}

ge::Status LLMDataDist::Impl::SwitchRole(const std::string &role, const std::map<std::string, std::string> &options) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(), ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(enable_switch_role_,
                         ge::LLM_FEATURE_NOT_ENABLED,
                         "Option %s not enabled",
                         LLM_OPTION_ENABLE_SWITCH_ROLE);
  LLM_CHK_BOOL_RET_STATUS(device_ids_.size() <= 1U, ge::LLM_FEATURE_NOT_ENABLED,
                         "Switch role is not supported when device num bigger than one.");
  LLM_CHK_STATUS_RET(llm_flow_service_->SwitchRole(role, options), "Failed to switch role");
  LLMEVENT("role switched from %s to %s", role_.c_str(), role.c_str());
  role_ = role;
  return ge::SUCCESS;
}

ge::Status LLMDataDist::Impl::SwapBlocks(const Cache &src, const Cache &dst, const uint64_t block_size,
                                         const uint32_t type,
                                         const std::vector<std::pair<int64_t, int64_t>> &block_mapping) {
  LLM_CHK_BOOL_RET_STATUS(is_initialized_.load(std::memory_order::memory_order_relaxed), ge::FAILED,
                         "Llm datadist of cluster:%lu is not initialized.", cluster_id_);
  LLM_CHK_BOOL_RET_STATUS(src.per_device_tensor_addrs.size() == device_ids_.size(), ge::LLM_PARAM_INVALID,
                         "Param invalid, per_device_tensor_addrs size:%zu of src cache should be equal to device num:%zu.",
                         src.per_device_tensor_addrs.size(), device_ids_.size());
  LLM_CHK_BOOL_RET_STATUS(dst.per_device_tensor_addrs.size() == device_ids_.size(), ge::LLM_PARAM_INVALID,
                         "Param invalid, per_device_tensor_addrs size:%zu of dst cache should be equal to device num:%zu.",
                         dst.per_device_tensor_addrs.size(), device_ids_.size());
  SwapBlockParam swap_param{block_size, type, block_mapping};
  return llm_flow_service_->SwapBlocks(src, dst, swap_param, device_ids_);
}

// LLMDataDist start
LLMDataDist::LLMDataDist(const uint64_t cluster_id) {
  LLMLOGD("Enter LLMDataDist construction.");
  impl_ = llm::MakeUnique<Impl>(cluster_id);
  if (impl_ == nullptr) {
    LLMLOGW("Create LlmEngineImpl failed");
  }
}

LLMDataDist::~LLMDataDist() = default;

ge::Status LLMDataDist::LLMDataDistInitialize(const std::map<ge::AscendString, ge::AscendString> &options) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes((impl_->LLMDataDistInitialize(options)));
}

void LLMDataDist::LLMDataDistFinalize() {
  if (impl_ != nullptr) {
    impl_->LLMDataDistFinalize();
  }
}

ge::Status LLMDataDist::CheckLinkStatus(uint64_t remote_cluster_id){
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes((impl_->CheckLinkStatus(remote_cluster_id)));
}

ge::Status LLMDataDist::LinkClusters(const std::vector<ClusterInfo> &clusters, std::vector<ge::Status> &rets,
                                     const int32_t timeout) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes((impl_->LinkClusters(clusters, rets, timeout)));
}

ge::Status LLMDataDist::UnlinkClusters(const std::vector<ClusterInfo> &clusters, std::vector<ge::Status> &rets,
                                       const int32_t timeout, bool force_flag) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes((impl_->UnlinkClusters(clusters, rets, timeout, force_flag)));
}

ge::Status LLMDataDist::AllocateCache(const CacheDesc &cache_desc,
                                      Cache &cache,
                                      const std::vector<CacheKey> &cache_keys) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->AllocateCache(cache_desc, cache, cache_keys));
}

ge::Status LLMDataDist::DeallocateCache(int64_t cache_id) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->DeallocateCache(cache_id));
}

ge::Status LLMDataDist::PullCache(int64_t cache_id, const CacheKey &cache_key, const PullCacheParam &pull_cache_param) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->PullCache(cache_id, cache_key, pull_cache_param));
}

ge::Status LLMDataDist::PullBlocks(int64_t cache_id, const CacheKey &cache_key, const PullCacheParam &pull_cache_param) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->PullBlocks(cache_id, cache_key, pull_cache_param));
}

ge::Status LLMDataDist::TransferCache(uint64_t task_id, const TransferCacheConfig &transfer_cache_config,
                                      const TransferBlockConfig &transfer_block_config) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->TransferCache(task_id, transfer_cache_config, transfer_block_config));
}

ge::Status LLMDataDist::CopyCache(const CacheEntry &src_cache_entry, const CacheEntry &dst_cache_entry,
                                  const CopyCacheParam &copy_cache_param, const std::vector<int32_t> &device_ids) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->CopyCache(src_cache_entry, dst_cache_entry, copy_cache_param, device_ids));
}

ge::Status LLMDataDist::CopyCache(const CopyCacheParam &copy_cache_param) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->CopyCache(copy_cache_param));
}

ge::Status LLMDataDist::RemoveCacheKey(const CacheKey &cache_key) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->RemoveCacheKey(cache_key));
}

ge::Status LLMDataDist::GetCacheTensors(int64_t cache_id,
                                        std::vector<ge::Tensor> &outputs,
                                        int32_t tensor_index) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->GetCacheTensors(cache_id, outputs, tensor_index));
}

ge::Status LLMDataDist::SwitchRole(const std::string &role, const std::map<std::string, std::string> &options) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->SwitchRole(role, options));
}

ge::Status LLMDataDist::SwapBlocks(const Cache &src, const Cache &dst, const uint64_t block_size, const uint32_t type,
                                   const std::vector<std::pair<int64_t, int64_t>> &block_mapping) {
  LLM_CHK_BOOL_RET_STATUS(impl_ != nullptr, ge::LLM_PARAM_INVALID, "impl is nullptr, check LLMDataDist construct");
  return TransRetToLlmCodes(impl_->SwapBlocks(src, dst, block_size, type, block_mapping));
}

bool LLMDataDist::IsInitialized() const {
  return impl_ != nullptr ? impl_->IsInitialized() : false;
}
}  // namespace llm
